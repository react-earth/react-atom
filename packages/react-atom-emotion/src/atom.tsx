import React, { CSSProperties, ElementType, forwardRef } from 'react';
import {
  AtomProps,
  ATOM_PSEUDO_CLASS_PROPS,
  DEFAULT_COMPONENT,
  parseAtomProps,
  PseudoClassStyle,
  Tokens,
} from 'react-atom-core';
import styled from '@emotion/styled';

const normalizeKey = (key: string) => key.replace(/([A-Z])/g, '-$1').toLowerCase();

const parseStyle = (style: CSSProperties) =>
  Object.entries(style)
    .map(([key, value]) => `${normalizeKey(key)}: ${value};`)
    .join('');

type AtomOptions = {
  defaultComponent?: ElementType;
};

export const atom = <T extends Tokens = Tokens>(tokens: T, options?: AtomOptions) => {
  const defaultComponent = options?.defaultComponent ?? DEFAULT_COMPONENT;

  return forwardRef<any, AtomProps<T>>((props, ref) => {
    const { as: Component = defaultComponent, ...rest } = props;
    const { style, pseudoClassStyle, htmlProps } = parseAtomProps(rest as AtomProps, tokens);
    return (
      <StyledAtom as={Component} $style={style} $pseudoClassStyle={pseudoClassStyle} ref={ref as any} {...htmlProps} />
    );
  });
};

const StyledAtom = styled.div<{ $style: CSSProperties; $pseudoClassStyle: PseudoClassStyle }>`
  ${({ $style }) => parseStyle($style)}
  ${({ $pseudoClassStyle }) => {
    return Object.entries($pseudoClassStyle)
      .sort(
        ([key1], [key2]) =>
          ATOM_PSEUDO_CLASS_PROPS.findIndex((key) => key === key2) -
          ATOM_PSEUDO_CLASS_PROPS.findIndex((key) => key === key1)
      )
      .map(
        ([key, style]) => `
          &:${normalizeKey(key)} {
            ${parseStyle(style)}
          }
        `
      )
      .join(' ');
  }}
`;
