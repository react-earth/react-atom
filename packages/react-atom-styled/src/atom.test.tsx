import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';
import { atom } from './atom';

const Atom = atom({
  color: {
    black: '#000000',
    white: '#FFFFFF',
    green: {
      50: '#F0FFF4',
      100: '#C6F6D5',
      200: '#9AE6B4',
    },
  },
  spacing: {
    1: '0.25rem',
    2: '0.5rem',
    3: '0.75rem',
    4: '1rem',
  },
  fontSize: {
    sm: '24rem',
    md: '28rem',
    lg: '32rem',
  },
});

describe('atom', () => {
  it('Atom component render correctly', () => {
    const { container } = render(
      <Atom as="span" c="green.50" width={500} height="4" fontSize="sm" $hover={{ c: 'white' }} $$after={{ w: '4' }} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
