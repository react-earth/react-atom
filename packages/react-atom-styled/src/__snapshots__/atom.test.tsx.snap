// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atom Atom component render correctly 1`] = `
.c0 {
  color: #F0FFF4;
  width: 500px;
  height: 1rem;
  font-size: 24rem;
}

.c0:hover {
  color: #FFFFFF;
}

.c0::after {
  width: 1rem;
}

<span
  class="c0"
/>
`;
